protocol/contracts/v2/safeGuards/ThresholdSafeGuard.sol

```diff
@@ -0,0 +1,236 @@
+      * the transaction once the minimum threshold of signers have approved
+     */
+    function safeTxExecutors() external view returns (address[] memory executors) {
+        return _safeTxExecutors.values();
+    }
+
+    /**
+      * @notice The required signatory threshold for a given contract and functionSignature
+     */
+    function getThreshold(address contractAddr, bytes4 fnSignature) public view returns (uint256) {
+        uint256 threshold = functionThresholds[contractAddr][fnSignature];
+        return threshold > 0 ? threshold : defaultSignaturesThreshold;
+    }
+
+    /**
+      * @notice The Safe will call this method once the signatories has been checked vs Safe's builtin threshold.
+      * This will revert:
+      *    1/ If the safeTxExecutor is not either a safe owner or pre-approved executor
+      *    2/ The number of signers does not meet a per contract/function defined threshold
+      *       (or a default threshold if not explicitly set per contract/function)
+      * 
+      * Exceptions:
+      *    a/ If the builtin Safe's threshold == 1, the number of signatories here isn't checked. 
+      *       So the proposer can execute immediately. 
+      *       This is to handle Tenderly transaction simulations via the dapp.
+      *    b/ If number of signers for the contract/function is less than the safe threshold - then no extra
+      *       signatory checks are done here, as it's already been done in the Safe.
+     */
+    function checkTransaction(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        // solhint-disable-next-line no-unused-vars
+        address payable refundReceiver,
+        bytes memory signatures,
+        address safeTxExecutor
+    ) external view {
+        // Circuit breaker
+        if (disableGuardChecks) return;
+
+        GnosisSafe safe = GnosisSafe(payable(msg.sender));
+
+        if (!(_safeTxExecutors.contains(safeTxExecutor) || safe.isOwner(safeTxExecutor))) {
+            revert InvalidExecutor();
+        }
+
+        uint256 threshold;
+        {
+            uint256 safeThreshold = safe.getThreshold();
+
+            // When Tenderly simulations run, the Safe's threshold is overriden to 1.
+            // So only check the signatures when the threshold is > 1
+            // WARNING: This means that we aren't doing any extra checks if the Safe's threshold == 1. It should always be > 1
+            // Note - if multiple signers have already signed, then the Safe dapp doesn't do this override. So it may still fail if
+            //        number of signers == 1 < x < dynamicThresholdRequirement
+            if (safeThreshold == 1) return;
+
+            // Perhaps in future there could be some custom decoding/approvals based on the arguments too
+            threshold = getThreshold(to, bytes4(data));
```

Comment:
if `data.length == 0`, defaultThreshold is set anyways. Are there executions for this case? Meaning a call to a contract/function-sig combination not allowed (while `data.length == 0`. Could also be a delegate call.
Might want to safe guard against such cases.

Reply:
The only thing I can think of where calldata would be null would be sending raw ETH? It's a good shout to test that explicitly too. 

Are you suggesting we should block delegatecall's? I think the safe website warns if that's the case, would need to make sure that the 'transaction batch' app still works in gnosis though as i think it uses delegatecall via the multicall contract

Reply:
I've added a test and updated such that it can work for an ETH transfer - ie empty bytes payload
https://github.com/TempleDAO/temple/commit/bb87aa50d75cc0496625cba3c682a15bbf2005bb

lmk your thoughts on `delegatecall`

Reply:
commit looks good. 
oh ok, then in that case it's okay to keep support for `delegateCall`